api/envoy/extensions/http/unified_header_validators/envoy_default/v3/unified_header_validator.proto

```diff
@@ -0,0 +1,97 @@
+option java_outer_classname = "UnifiedHeaderValidatorProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/http/unified_header_validators/envoy_default/v3;envoy_defaultv3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Envoy's default Unified Header Validator config]
+
+// This extension validates that HTTP request and response headers are well formed according to respective RFCs.
+// 1. H/1 header map validity according to `RFC 7230 section 3.2 <https://tools.ietf.org/html/rfc7230#section-3.2>`_
+// 2. Syntax of H/1 request target URI and response status
+// 3. H/2 header map validity according to `RFC 7540 section 8.1.2 <https://tools.ietf.org/html/rfc7540#section-8.1.2>`_
+// 4. Syntax of H/2 pseudo headers
+// 5. Syntax of Content-Length and Transfer-Encoding
+// 6. Validation of H/1 requests with both Content-Length and Transfer-Encoding headers
+// 7. Normalization of the URI path.
+//
+message UnifiedHeaderValidatorConfig {
+  // Determines the action for request that contain %2F, %2f, %5C or %5c sequences in the URI path.
+  // This operation occurs before URL normalization and the merge slashes transformations if they were enabled.
+  enum PathWithEscapedSlashesAction {
+    // Default behavior specific to implementation (i.e. Envoy) of this configuration option.
+    // Envoy, by default, takes the KEEP_UNCHANGED action.
+    // NOTE: the implementation may change the default behavior at-will.
+    IMPLEMENTATION_SPECIFIC_DEFAULT = 0;
+
+    // Keep escaped slashes.
+    KEEP_UNCHANGED = 1;
+
+    // Reject client request with the 400 status. gRPC requests will be rejected with the INTERNAL (13) error code.
+    // The "httpN.downstream_rq_failed_path_normalization" counter is incremented for each rejected request.
+    REJECT_REQUEST = 2;
+
+    // Unescape %2F and %5C sequences and redirect request to the new path if these sequences were present.
+    // Redirect occurs after path normalization and merge slashes transformations if they were configured.
+    // NOTE: gRPC requests will be rejected with the INTERNAL (13) error code.
+    // This option minimizes possibility of path confusion exploits by forcing request with unescaped slashes to
+    // traverse all parties: downstream client, intermediate proxies, Envoy and upstream server.
+    // The "httpN.downstream_rq_redirected_with_normalized_path" counter is incremented for each
+    // redirected request.
+    UNESCAPE_AND_REDIRECT = 3;
+
+    // Unescape %2F and %5C sequences.
+    // Note: this option should not be enabled if intermediaries perform path based access control as
+    // it may lead to path confusion vulnerabilities.
+    UNESCAPE_AND_FORWARD = 4;
+  }
+
+  message Http1ProtocolOptions {
+    // Allows Envoy to process HTTP/1 requests/responses with both `Content-Length` and `Transfer-Encoding`
```

Comment:
as long as we're adding legacy mode, do you think we should have a boolean to restrict HTTP/1.1 methods down to what http_parser did, or wait and see if anyone wants it?

Reply:
yes, makes sense. I will be adding other options as well from HCM and config.core.v3.Http1ProtocolOptions protos as we move along.